src/OpenSage.Mods.Generals/Gui/GeneralsControlBar.cs

```diff
@@ -313,8 +322,39 @@ public override void Update(Player player, GeneralsControlBar controlBar)
 
                 var unitSelectedControl = controlBar._right.Controls.FindControl("ControlBar.wnd:WinUnitSelected");
 
+                var productionQueueWindow = controlBar._right.Controls.FindControl("ControlBar.wnd:ProductionQueueWindow");
+                productionQueueWindow.Visible = unit.IsProducing;
+
+                var queue = unit.ProductionQueue;
+
+                for(int pos = 0; pos < 9; pos++)
+                {
+                    var queueButton = productionQueueWindow.Controls.FindControl($"ControlBar.wnd:ButtonQueue0{pos+1}");
+                    Image img = null;
+                    if (queue.Count > pos)
+                    {
+                        var job = queue[pos];
+                        if (queueButton != null && job != null)
+                        {
+                            //quick and dirty progress indicator. needs to be remade to show the clock-like overlay
+                            queueButton.Opacity = (1.0f - job.Progress);
+
+                            img = controlBar._contentManager.WndImageLoader.CreateNormalImage(job.objectDefinition.SelectPortrait);
+                            queueButton.SystemCallback = (Control control, WndWindowMessage message, ControlCallbackContext context) =>
+                            {
+                                unit.CancelProduction(pos);
```

Comment:
There is an issue with this line: https://stackoverflow.com/questions/235455/access-to-modified-closure

Basically, the `pos` variable is modified by the loop, so the callback doesn't work quite as expected. Maybe something like this will fix it:

```csharp
var posCopy = pos;
queueButton.SystemCallback = (control, message, context) => unit.CancelProduction(posCopy);
```

Reply:
done